File: apps/api/services/director/overrides.py
"""Director Engine Phase 6 — non-destructive timeline overrides (Python mirror of overrides.ts)."""
from __future__ import annotations

import copy
from typing import Any

from services.director.resolve_broll import reroll_broll_with_pexels

LAYER_DEPTH: dict[str, int] = {
    "metric_ticker": 52,
    "animated_title": 48,
    "lower_third": 55,
    "kinetic_karaoke": 58,
    "strategy_funnel": 35,
    "device_mockup": 28,
}


def _layer_depth(component_id: str) -> int:
    return LAYER_DEPTH.get(component_id, 50)


def _infer_component_id(trigger: dict[str, Any]) -> str:
    meta = trigger.get("metadata") or {}
    comp = meta.get("componentId")
    if isinstance(comp, str) and comp:
        return comp
    return "animated_title"


def delete_timeline_entry(timeline: dict[str, Any], entry_id: str) -> dict[str, Any]:
    next_tl = copy.deepcopy(timeline)
    tracks = next_tl.setdefault("tracks", {})
    tracks["motionGraphics"] = [
        e for e in tracks.get("motionGraphics", []) if e.get("id") != entry_id
    ]
    tracks["broll"] = [e for e in tracks.get("broll", []) if e.get("id") != entry_id]

    triggers = []
    for t in next_tl.get("triggers", []):
        if t.get("resultingEntryId") == entry_id:
            triggers.append({**t, "status": "suppressed", "resultingEntryId": None})
        else:
            triggers.append(t)
    next_tl["triggers"] = triggers
    return next_tl


def promote_trigger(
    timeline: dict[str, Any],
    trigger_id: str,
    component_id: str | None = None,
) -> dict[str, Any]:
    next_tl = copy.deepcopy(timeline)
    trigger = next((t for t in next_tl.get("triggers", []) if t.get("id") == trigger_id), None)
    if not trigger or trigger.get("status") != "suppressed":
        return timeline

    fps = float(next_tl.get("fps") or 30)
    comp_id = component_id or _infer_component_id(trigger)
    start_frame = round(float(trigger.get("transcriptStart", 0)) * fps)
    end_frame = max(start_frame + 1, round(float(trigger.get("transcriptEnd", 0)) * fps))
    entry_id = f"entry-{trigger_id}-promoted"

    tracks = next_tl.setdefault("tracks", {})
    mg = list(tracks.get("motionGraphics", []))
    mg.append(
        {
            "id": entry_id,
            "componentId": comp_id,
            "startFrame": start_frame,
            "durationInFrames": end_frame - start_frame,
            "layerDepth": _layer_depth(comp_id),
            "props": (trigger.get("metadata") or {}).get("props") or {},
            "triggerId": trigger_id,
        }
    )
    tracks["motionGraphics"] = mg

    next_tl["triggers"] = [
        (
            {**t, "status": "realized", "resultingEntryId": entry_id}
            if t.get("id") == trigger_id
            else t
        )
        for t in next_tl.get("triggers", [])
    ]
    return next_tl


def swap_timeline_component(
    timeline: dict[str, Any],
    entry_id: str,
    new_component_id: str,
    props: dict[str, Any] | None = None,
) -> dict[str, Any]:
    next_tl = copy.deepcopy(timeline)
    tracks = next_tl.setdefault("tracks", {})
    mg = []
    for entry in tracks.get("motionGraphics", []):
        if entry.get("id") != entry_id:
            mg.append(entry)
            continue
        mg.append(
            {
                **entry,
                "componentId": new_component_id,
                "layerDepth": _layer_depth(new_component_id),
                "props": props if props is not None else entry.get("props", {}),
            }
        )
    tracks["motionGraphics"] = mg
    return next_tl


def reroll_broll_entry(
    timeline: dict[str, Any],
    entry_id: str,
    search_query: str,
    *,
    content_type: str = "podcast",
) -> dict[str, Any]:
    return reroll_broll_with_pexels(
        timeline,
        entry_id,
        search_query,
        content_type=content_type,
    )

File: remotion-service/src/lib/director/overrides.ts
import type { DirectorTimeline, TriggerLogEntry } from "@types/timeline";
import { layerDepthForComponent } from "./layerRegistry";

/** Remove a realized entry and mark its trigger suppressed (Phase 6). */
export function deleteTimelineEntry(
  timeline: DirectorTimeline,
  entryId: string,
): DirectorTimeline {
  const next = structuredClone(timeline);
  next.tracks.motionGraphics = next.tracks.motionGraphics.filter((e) => e.id !== entryId);
  next.tracks.broll = next.tracks.broll.filter((e) => e.id !== entryId);

  next.triggers = next.triggers.map((t) => {
    if (t.resultingEntryId !== entryId) return t;
    return { ...t, status: "suppressed" as const, resultingEntryId: undefined };
  });

  return next;
}

/** Promote a suppressed trigger back onto the Timeline. */
export function promoteTrigger(
  timeline: DirectorTimeline,
  triggerId: string,
  componentId?: string,
): DirectorTimeline {
  const next = structuredClone(timeline);
  const trigger = next.triggers.find((t) => t.id === triggerId);
  if (!trigger || trigger.status !== "suppressed") return timeline;

  const compId = componentId ?? inferComponentId(trigger);
  const startFrame = Math.round(trigger.transcriptStart * next.fps);
  const endFrame = Math.max(startFrame + 1, Math.round(trigger.transcriptEnd * next.fps));
  const entryId = `entry-${triggerId}-promoted`;

  next.tracks.motionGraphics.push({
    id: entryId,
    componentId: compId,
    startFrame,
    durationInFrames: endFrame - startFrame,
    layerDepth: layerDepthForComponent(compId),
    props: (trigger.metadata?.props as Record<string, unknown>) ?? {},
    triggerId,
  });

  next.triggers = next.triggers.map((t) =>
    t.id === triggerId
      ? { ...t, status: "realized" as const, resultingEntryId: entryId }
      : t,
  );

  return next;
}

/** Swap the component for an existing motion graphics slot. */
export function swapTimelineComponent(
  timeline: DirectorTimeline,
  entryId: string,
  newComponentId: string,
  props?: Record<string, unknown>,
): DirectorTimeline {
  const next = structuredClone(timeline);
  next.tracks.motionGraphics = next.tracks.motionGraphics.map((e) => {
    if (e.id !== entryId) return e;
    return {
      ...e,
      componentId: newComponentId,
      layerDepth: layerDepthForComponent(newComponentId),
      props: props ?? e.props,
    };
  });
  return next;
}

/** Re-roll a single B-roll slot with a new search query. */
export function rerollBrollEntry(
  timeline: DirectorTimeline,
  entryId: string,
  searchQuery: string,
): DirectorTimeline {
  const next = structuredClone(timeline);
  next.tracks.broll = next.tracks.broll.map((e) =>
    e.id === entryId ? { ...e, assetUrl: "", searchQuery } : e,
  );
  return next;
}

function inferComponentId(trigger: TriggerLogEntry): string {
  const fromMeta = trigger.metadata?.componentId;
  if (typeof fromMeta === "string") return fromMeta;
  return "animated_title";
}

export function listSuppressedTriggers(timeline: DirectorTimeline): TriggerLogEntry[] {
  return timeline.triggers.filter((t) => t.status === "suppressed");
}
